plugins/inputs/slurm/slurm.go

```diff
@@ -0,0 +1,508 @@
+			records["architecture"] = *strPtr
+		}
+
+		acc.AddFields("slurm_nodes", records, tags)
+	}
+}
+
+func (s *Slurm) gatherPartitionsMetrics(acc telegraf.Accumulator,
+	partitions []goslurm.V0038Partition) {
+	var (
+		int32Ptr *int32
+		strPtr   *string
+		ok       bool
+	)
+
+	for _, partition := range partitions {
+		records := make(map[string]interface{})
+		tags := make(map[string]string)
+
+		tags["source"] = s.baseURL.Hostname()
+		if strPtr, ok = partition.GetNameOk(); ok {
+			tags["name"] = *strPtr
+		}
+
+		if strPtr, ok = partition.GetStateOk(); ok {
+			records["state"] = *strPtr
+		}
+		if int32Ptr, ok = partition.GetTotalCpusOk(); ok {
+			records["total_cpu"] = *int32Ptr
+		}
+		if int32Ptr, ok = partition.GetTotalNodesOk(); ok {
+			records["total_nodes"] = *int32Ptr
+		}
+		if strPtr, ok = partition.GetNodesOk(); ok {
+			records["nodes"] = *strPtr
+		}
+		if strPtr, ok = partition.GetTresOk(); ok {
+			for k, v := range s.parseTres(*strPtr) {
+				records["tres_"+k] = v
+			}
+		}
+
+		acc.AddFields("slurm_partitions", records, tags)
+	}
+}
+
+func (s *Slurm) gatherReservationsMetrics(acc telegraf.Accumulator,
+	reservations []goslurm.V0038Reservation) {
+	var (
+		int32Ptr *int32
+		strPtr   *string
+		ok       bool
+	)
+	for _, reservation := range reservations {
+		records := make(map[string]interface{})
+		tags := make(map[string]string)
+
+		tags["source"] = s.baseURL.Hostname()
+		if strPtr, ok = reservation.GetNameOk(); ok {
+			tags["name"] = *strPtr
+		}
+
+		if int32Ptr, ok = reservation.GetCoreCountOk(); ok {
+			records["core_count"] = *int32Ptr
+		}
+		if int32Ptr, ok = reservation.GetCoreSpecCntOk(); ok {
+			records["core_spec_count"] = *int32Ptr
+		}
+		if strPtr, ok = reservation.GetGroupsOk(); ok {
+			records["groups"] = *strPtr
+		}
+		if strPtr, ok = reservation.GetUsersOk(); ok {
+			records["users"] = *strPtr
+		}
+		if int32Ptr, ok = reservation.GetStartTimeOk(); ok {
+			records["start_time"] = *int32Ptr
+		}
+		if strPtr, ok = reservation.GetPartitionOk(); ok {
+			records["partition"] = *strPtr
+		}
+		if strPtr, ok = reservation.GetAccountsOk(); ok {
+			records["accounts"] = *strPtr
+		}
+		if int32Ptr, ok = reservation.GetNodeCountOk(); ok {
+			records["node_count"] = *int32Ptr
+		}
+		if strPtr, ok = reservation.GetNodeListOk(); ok {
+			records["node_list"] = *strPtr
+		}
+
+		acc.AddFields("slurm_reservations", records, tags)
+	}
+}
+
+func (s *Slurm) Gather(acc telegraf.Accumulator) (err error) {
+	auth := context.WithValue(
+		context.Background(),
+		goslurm.ContextAPIKeys,
+		map[string]goslurm.APIKey{
+			"user":  {Key: s.Username},
+			"token": {Key: s.Token},
+		},
+	)
+
+	if !s.endpointMap["diag"] {
+		diagResp, respRaw, err := s.client.SlurmAPI.SlurmV0038Diag(auth).Execute()
+		if err != nil {
+			return fmt.Errorf("error getting diag: %w", err)
+		}
+		defer respRaw.Body.Close()
+		if diag, ok := diagResp.GetStatisticsOk(); ok {
+			s.gatherDiagMetrics(acc, diag)
+		}
```

Comment:
I really would use
```suggestion
		if diag, ok := diagResp.GetStatisticsOk(); ok {
			s.gatherDiagMetrics(acc, diag)
		}
		respRaw.Body.Close()
```
as the deferral will keep the body until _all_ metrics are gathered instead of closing it after use... Same for the other statements below.

Reply:
Given only 5 HTTP requests were done I didn't think twice about it but yeah, closing the body ASAP is the better practice for sure.

Anyway, I added the changes on ab4d126a71266b1a1839d52d9274d16f39d25777. Thanks for the tip!